UofT-Frosh-Orientation/orientation-website
# 621 confirming emails when creating accounts

## **Contribution:**

### **Issue:**
Closes #621 

### **Accomplishments:**
- Edited Sign Up page to tell users to check inbox for a verification link.
- Added 'confirmed' flag in `UserModel.js` which checks if they verified their account through email token or not
- Sends email token upon signing up using EmailServices.js. Instead of making a new function in `UserServices.js,` I tried to  update createUser so that an email is sent after creating an account. 
- Updated `userRoutes.js` with a route to confirmUser in `UserController.js`. This employs the validateEmailConfirmationToken function in `UserServices.js` to check the token and change the flag accordingly.
- Included another condition to the login function in `UserController.js` so that unconfirmed users should not be able to log in
- Created new page called `EmailConfirmed.jsx` to display status after the email token is sent. This should show a 'Perfect' or 'Oh no' message and have a button to let users log in again.

### **Visuals:** 
(I'll update once Docker works again)
**Updated Sign up Page:**
![image](https://github.com/UofT-Frosh-Orientation/orientation-website/assets/109243682/e253bf66-a642-4773-b5b3-232595edf05c)
**Verification Complete:**
![image](https://github.com/UofT-Frosh-Orientation/orientation-website/assets/109243682/7977a715-5b74-4a3a-bc3c-481b75642a7d)
**Error message for incorrect token:** 
![image](https://github.com/UofT-Frosh-Orientation/orientation-website/assets/109243682/ca072edf-cb41-40e6-839a-bee08bfd3eb4)
**Simple Email with URL**
![image](https://github.com/UofT-Frosh-Orientation/orientation-website/assets/109243682/31c22a6c-626e-40c1-8388-746783835b50)


## **Details:**

### **Expected Behavior:**
- You should be able to sign up normally. The message after submitting the sign up form should tell you to check your inbox.
- An email with a URL should be sent that you can click to verify your account.
- Your account should then be confirmed if done correctly and you will be prompted to log in again. Otherwise, you will get an 'Oh no' message and have a button to try signing up again.

### **Testing Steps:**
1.  Set `readyForFrosh = true`
2. Run the backend using `docker-compose --up build`
3. Head to the login page --> Create account
4. Put in valid email address and info, check your inbox, click the link, and see if your account was confirmed correctly. You can also put invalid ones and check if you are able to login, too.

### **Complications:**
- Had some problems with committing, I need to check if it added the right files. Also, I did use `git config --unset core.hooksPath` to 'fix' the issue, not sure if this interferes with anything in dev.
- Right after signup, the user is automatically logged in. If they go through with verification they have to log in again, but not sure how to stop the initial one
- Wasn't able to use a separete .env token, so I just used the reset one to test. Let me know if there's a workaround.
- I have no clue how to use the template email so I just used a simple one.

### **Resources:**
- For my committing issues: [(https://github.com/typicode/husky/issues/1038)]
